### PR TITLE
Checkout callbacks

### PR DESCRIPTION
**Pull request by netnose**
- Added CheckoutNotifyType mapping git_checkout_notify_t
- Added CheckoutOpts.NotifyFlags of type CheckoutNotifyType
- Added CheckoutNotifyCallback mapping git_checkout_notify_cb
- Added CheckoutProgressCallback mapping git_checkout_progress_cb

### CODE REVIEW

File: wrapper.c [hunk 1]
```diff
@@ -10,6 +10,12 @@ void _go_git_populate_remote_cb(git_clone_options *opts)
 	opts->remote_cb = (git_remote_create_cb)remoteCreateCallback;
 }
 
+void _go_git_populate_checkout_cb(git_checkout_options *opts)
```
Comment (carlosmn) [inline wrapper.c]: This doesn't just populate the checkout, so it should reflect that in its name.

File: checkout.go [hunk 2]
```diff
@@ -83,6 +142,16 @@ func populateCheckoutOpts(ptr *C.git_checkout_options, opts *CheckoutOpts) *C.gi
...
+	ptr.notify_flags = C.uint(opts.NotifyFlags)
+	if opts.NotifyCallback != nil || opts.ProgressCallback != nil {
+		C._go_git_populate_checkout_cb(ptr)
+	}
+	if opts.NotifyCallback != nil {
+		ptr.notify_payload = pointerHandles.Track(*opts)
+	}
+	if opts.ProgressCallback != nil {
+		ptr.progress_payload = pointerHandles.Track(*opts)
+	}
```
Comment (carlosmn) [inline checkout.go]: Why track the same pointer twice?

Comment (carlosmn) [inline checkout.go]: And aren't we asking `Track()` here to track the structure itself? We should be asking it to track the _pointer_ which would be `opts` itself.

File: checkout.go [hunk 3]
```diff
@@ -70,6 +94,41 @@ func (opts *CheckoutOpts) toC() *C.git_checkout_options {
...
+		return 0
+	}
+	return int(opts.NotifyCallback(CheckoutNotifyType(why), path, baseline, target, workdir))
+}
+
+//export checkoutProgressCallback
+func checkoutProgressCallback(path *C.char, completed_steps, total_steps C.size_t, data unsafe.Pointer) int {
+	if data == nil {
+		return 0
+	}
```
Comment (carlosmn) [inline checkout.go]: When would we be registering the progress (or notify) callback but not set the payload pointer? Wouldn't hitting this mean we're silently hiding a bug in the bindings?

File: wrapper.c (hunk 1 above)
Comment (netnose) [inline wrapper.c]: What do you mean?
Something like _go_git_populate_checkout_cbs or _go_git_populate_checkout_notify_and_progress_cbs?

File: checkout.go (hunk 3 above)
Comment (netnose) [inline checkout.go]: We'll always set the payload pointer but this function is called by external code and checking the arguments is always a good practice.

Comment (carlosmn) [inline checkout.go]: It's only called as specified by our code, and if we're not passed `data` we have no reason to suspect anything else that gets passed in is valid. Why not check `path`? We're already assuming it's the right struct later on so we should either accept that we're asserting here, or return an error indicating that something has gone very wrong somewhere.

File: wrapper.c (hunk 1 above)
Comment (carlosmn) [inline wrapper.c]: Nevermind, I'm not sure what I was thinking.

**Pull request merged**